### PR TITLE
[mlir][sparse] Print new syntax

### PR DESCRIPTION
**Pull request by yinying-lisa-li**
Printing changes from `#sparse_tensor.encoding<{ lvlTypes = [ "compressed" ] }>` to `map = (d0) -> (d0 : compressed)`. Level properties, ELL and slice are also supported.

Some check tests are modified to support this change. For instance, `%[[S:.*]]:` is changed to `%[[S:[a-zA-Z0-9_]*]]:`, because the new syntax contains ":" and `.*` will greedily match sparse encoding instead of just the variable name.

### CODE REVIEW

File: mlir/test/Dialect/SparseTensor/codegen.mlir [hunk 1]
```diff
@@ -533,7 +533,7 @@ func.func @sparse_compression(%tensor: tensor<8x8xf64, #CSR>,
 //       CHECK:     %[[A13:.*]]:4 = scf.for %[[A14:.*]] = %[[A11]] to %[[A7]] step %[[A12]] iter_args(%[[A15:.*]] = %[[A0]], %[[A16:.*]] = %[[A1]], %[[A17:.*]] = %[[A2]], %[[A18:.*]] = %[[A3]]) -> (memref<?xindex>, memref<?xindex>, memref<?xf64>, !sparse_tensor.storage_specifier
 //       CHECK:       %[[A19:.*]] = memref.load %[[A6]]{{\[}}%[[A14]]] : memref<?xindex>
 //       CHECK:       %[[A20:.*]] = memref.load %[[A4]]{{\[}}%[[A19]]] : memref<?xf64>
-//       CHECK:       %[[A21:.*]]:4 = func.call @_insert_dense_compressed_no_8_8_f64_0_0(%[[A15]], %[[A16]], %[[A17]], %[[A18]], %[[A8]], %[[A19]], %[[A20]]) : (memref<?xindex>, memref<?xindex>, memref<?xf64>, !sparse_tensor.storage_specifier
+//       CHECK:       %[[A21:.*]]:4 = func.call @"_insert_dense_compressed(nonordered)_8_8_f64_0_0"(%[[A15]], %[[A16]], %[[A17]], %[[A18]], %[[A8]], %[[A19]], %[[A20]]) : (memref<?xindex>, memref<?xindex>, memref<?xf64>, !sparse_tensor.storage_specifier
```
Comment (aartbik) [inline mlir/test/Dialect/SparseTensor/codegen.mlir:536]: here and elsewhere, let's use compressed(nonordered) in the printing of type, but not in the generation for methods names (replace ( ) with _ perhaps, so we keep valid C identifiiers)

Comment (yinying-lisa-li) [reply mlir/test/Dialect/SparseTensor/codegen.mlir]: Done!

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 2]
```diff
@@ -586,30 +586,56 @@ Attribute SparseTensorEncodingAttr::parse(AsmParser &parser, Type type) {
...
+  for (unsigned i = 0; i < map.getNumResults() - 1; ++i) {
+    map.getResult(i).print(printer.getStream());
+    printer << " : " << toMLIRString(lvlTypes[i]) << ", ";
+  }
+  if (map.getNumResults() >= 1) {
+    auto lastIndex = map.getNumResults() - 1;
+    map.getResult(lastIndex).print(printer.getStream());
+    printer << " : " << toMLIRString(lvlTypes[lastIndex]);
+  }
+  printer << ')';
```
Comment (PeimingLiu) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:633]: I would suggest you break these into smaller functions.

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 3]
```diff
@@ -586,30 +586,56 @@ Attribute SparseTensorEncodingAttr::parse(AsmParser &parser, Type type) {
...
-    printer << "\"" << toMLIRString(dlt) << "\"";
-  });
-  printer << " ]";
+  auto map = static_cast<AffineMap>(getDimToLvl());
+  auto lvlTypes = getLvlTypes();
+  // Empty affine map indicates identity map
+  if (!map) {
+    map = AffineMap::getMultiDimIdentityMap(getLvlTypes().size(), getContext());
+  }
+  // Modified version of AsmPrinter::Impl::printAffineMap.
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:595]: I would remove this. This is diverged sufficiently to no longer refer to printAffineMap

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorCodegen.cpp [hunk 4]
```diff
@@ -472,8 +472,11 @@ class SparseInsertGenerator
     llvm::raw_svector_ostream nameOstream(nameBuffer);
     nameOstream << kInsertFuncNamePrefix;
     const Level lvlRank = stt.getLvlRank();
-    for (Level l = 0; l < lvlRank; l++)
-      nameOstream << toMLIRString(stt.getLvlType(l)) << "_";
+    for (Level l = 0; l < lvlRank; l++) {
+      std::string lvlType = toMLIRString(stt.getLvlType(l));
+      replaceWithUnderscore(lvlType);
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorCodegen.cpp:477]: std::string::replace ?

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 5]
```diff
@@ -586,31 +586,67 @@ Attribute SparseTensorEncodingAttr::parse(AsmParser &parser, Type type) {
...
-                            // skip mnemonic.
-                            attr.print(printer);
-                          });
-    printer << " ]";
+  printer << " }>";
+}
+
+void SparseTensorEncodingAttr::printSymbol(AffineMap &map,
+                                           AsmPrinter &printer) const {
+  if (map.getNumSymbols() != 0) {
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:611]: Note that LLVM slightly prefers

if (map.getNumSymbols() == 0)
  return;
... rest of code without indentation

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 6]
```diff
@@ -586,31 +586,67 @@ Attribute SparseTensorEncodingAttr::parse(AsmParser &parser, Type type) {
...
+    printer << ']';
   }
+}
 
-  printer << " }>";
+void SparseTensorEncodingAttr::printDimension(
+    AffineMap &map, AsmPrinter &printer,
+    ArrayRef<SparseTensorDimSliceAttr> dimSlices) const {
+  if (!dimSlices.empty()) {
+    for (unsigned i = 0; i < map.getNumDims() - 1; ++i)
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:625]: make invariant code more obvious with the following idiom

for (unsigned i = 0, n = map.getNumDims() - 1; i < n; i++)

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 7]
```diff
@@ -586,31 +586,67 @@ Attribute SparseTensorEncodingAttr::parse(AsmParser &parser, Type type) {
...
+      printer << 'd' << i << " : " << dimSlices[i] << ", ";
+    if (map.getNumDims() >= 1)
+      printer << 'd' << map.getNumDims() - 1 << " : "
+              << dimSlices[map.getNumDims() - 1];
+  } else {
+    for (unsigned i = 0; i < map.getNumDims() - 1; ++i)
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:631]: same

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 8]
```diff
@@ -586,31 +586,67 @@ Attribute SparseTensorEncodingAttr::parse(AsmParser &parser, Type type) {
...
+      printer << 'd' << i << ", ";
+    if (map.getNumDims() >= 1)
+      printer << 'd' << map.getNumDims() - 1;
+  }
+}
+
+void SparseTensorEncodingAttr::printLevel(
+    AffineMap &map, AsmPrinter &printer,
+    ArrayRef<DimLevelType> lvlTypes) const {
+  for (unsigned i = 0; i < map.getNumResults() - 1; ++i) {
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:641]: same

File: mlir/test/Dialect/SparseTensor/sparse_reshape.mlir [hunk 9]
```diff
@@ -16,7 +16,7 @@
 //      CHECK-ROUND:  return %[[E]] : tensor<10x10xf64, #sparse_tensor.encoding<{{{.*}}}>>
 //
 // CHECK-LABEL:   func.func @sparse_expand(
-// CHECK-SAME:    %[[S:.*]]:
+// CHECK-SAME:    %[[S:[a-zA-Z0-9_]*]]:
```
Comment (aartbik) [inline mlir/test/Dialect/SparseTensor/sparse_reshape.mlir:19]: why this change, or better, shouldn't we match the parameter part of the argument also?

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorAttrDefs.td [hunk 10]
```diff
@@ -422,6 +422,14 @@ def SparseTensorEncodingAttr : SparseTensor_Attr<"SparseTensorEncoding",
     std::optional<uint64_t> getStaticLvlSliceOffset(::mlir::sparse_tensor::Level lvl) const;
     std::optional<uint64_t> getStaticLvlSliceSize(::mlir::sparse_tensor::Level lvl) const;
     std::optional<uint64_t> getStaticLvlSliceStride(::mlir::sparse_tensor::Level lvl) const;
+
+    //
+    // Printing methods.
+    //
+
+    void printSymbol(AffineMap &map, AsmPrinter &printer) const;
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorAttrDefs.td:430]: one last nit: can we make these plural, as in printSymbols, printDimensions, printLevels
other than that go to go!

**Review by aartbik — APPROVED** [review]

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorAttrDefs.td (hunk 10 above)
Comment (yinying-lisa-li) [reply mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorAttrDefs.td]: Sounds good. Plural form fits the other functions names like getLvlTypes.
